File: bitswan-workspace-dashboard/server/src/routes/templates.ts
import type { FastifyInstance } from 'fastify';
import { isValidBpId, isValidWorktreeName } from '../services/workspace.js';
import type { GitopsClient } from '../services/gitops.js';

export interface TemplateRoutesOptions {
  gitops: GitopsClient | null;
}

/**
 * Template discovery + automation creation. Both endpoints proxy to gitops,
 * which owns the bind-mount on `/workspace/examples` and the only write path
 * into `/workspace-repo`. The dashboard server's role here is auth-boundary
 * and request validation.
 */
export function registerTemplateRoutes(
  app: FastifyInstance,
  { gitops }: TemplateRoutesOptions,
): void {
  app.get('/api/templates', async (_req, reply) => {
    reply.header('Cache-Control', 'no-store');
    if (!gitops) {
      return reply.code(503).send({ error: 'gitops not configured' });
    }
    try {
      const r = await gitops.getTemplates();
      if (!r.ok) {
        return reply
          .code(r.status >= 400 && r.status < 500 ? r.status : 502)
          .send({ error: 'gitops error', status: r.status, body: r.body });
      }
      return r.body;
    } catch (err) {
      app.log.warn({ err }, 'templates fetch failed');
      return reply.code(502).send({ error: 'gitops unreachable' });
    }
  });

  app.post<{
    Body: {
      template_id?: string;
      group_id?: string;
      name?: string;
      bp?: string;
      worktree?: string;
    };
  }>('/api/automations/from-template', async (req, reply) => {
    reply.header('Cache-Control', 'no-store');
    if (!gitops) {
      return reply.code(503).send({ error: 'gitops not configured' });
    }
    const { template_id, group_id, name, bp, worktree } = req.body ?? {};
    if (!bp || !isValidBpId(bp)) {
      return reply.code(400).send({ error: 'invalid bp' });
    }
    if (worktree !== undefined && !isValidWorktreeName(worktree)) {
      return reply.code(400).send({ error: 'invalid worktree' });
    }
    if (!template_id && !group_id) {
      return reply.code(400).send({ error: 'template_id or group_id required' });
    }
    if (template_id && group_id) {
      return reply
        .code(400)
        .send({ error: 'template_id and group_id are mutually exclusive' });
    }
    try {
      const r = await gitops.createAutomationFromTemplate({
        ...(template_id ? { template_id } : {}),
        ...(group_id ? { group_id } : {}),
        ...(name !== undefined ? { name } : {}),
        bp,
        ...(worktree ? { worktree } : {}),
      });
      if (!r.ok) {
        return reply
          .code(r.status >= 400 && r.status < 500 ? r.status : 502)
          .send({ error: 'gitops error', status: r.status, body: r.body });
      }
      return r.body;
    } catch (err) {
      app.log.warn({ err, body: req.body }, 'create-from-template failed');
      return reply.code(502).send({ error: 'gitops unreachable' });
    }
  });

  // Stage 1.5 scaffolding: add a frontend / worker / rename within a BP. Each
  // proxies straight to the matching gitops endpoint; the dashboard's job here
  // is the auth boundary + request validation (gitops owns the write path).
  const relay = async (
    reply: import('fastify').FastifyReply,
    label: string,
    call: () => Promise<{ ok: boolean; status: number; body: unknown }>,
  ) => {
    reply.header('Cache-Control', 'no-store');
    if (!gitops) {
      return reply.code(503).send({ error: 'gitops not configured' });
    }
    try {
      const r = await call();
      if (!r.ok) {
        return reply
          .code(r.status >= 400 && r.status < 500 ? r.status : 502)
          .send({ error: 'gitops error', status: r.status, body: r.body });
      }
      return r.body;
    } catch (err) {
      app.log.warn({ err }, `${label} failed`);
      return reply.code(502).send({ error: 'gitops unreachable' });
    }
  };

  app.post<{
    Body: { bp?: string; name?: string; worktree?: string };
  }>('/api/automations/frontend', async (req, reply) => {
    const { bp, name, worktree } = req.body ?? {};
    if (!bp || !isValidBpId(bp)) return reply.code(400).send({ error: 'invalid bp' });
    if (!name) return reply.code(400).send({ error: 'name required' });
    if (worktree !== undefined && !isValidWorktreeName(worktree)) {
      return reply.code(400).send({ error: 'invalid worktree' });
    }
    return relay(reply, 'add-frontend', () =>
      gitops!.addFrontend({ bp, name, ...(worktree ? { worktree } : {}) }),
    );
  });

  app.post<{
    Body: { bp?: string; name?: string; type?: string; worktree?: string };
  }>('/api/automations/worker', async (req, reply) => {
    const { bp, name, type, worktree } = req.body ?? {};
    if (!bp || !isValidBpId(bp)) return reply.code(400).send({ error: 'invalid bp' });
    if (!name) return reply.code(400).send({ error: 'name required' });
    if (!type) return reply.code(400).send({ error: 'type required' });
    if (worktree !== undefined && !isValidWorktreeName(worktree)) {
      return reply.code(400).send({ error: 'invalid worktree' });
    }
    return relay(reply, 'add-worker', () =>
      gitops!.addWorker({ bp, name, type, ...(worktree ? { worktree } : {}) }),
    );
  });

  app.post<{
    Body: { bp?: string; old_name?: string; new_name?: string; worktree?: string };
  }>('/api/automations/rename', async (req, reply) => {
    const { bp, old_name, new_name, worktree } = req.body ?? {};
    if (!bp || !isValidBpId(bp)) return reply.code(400).send({ error: 'invalid bp' });
    if (!old_name || !new_name) {
      return reply.code(400).send({ error: 'old_name and new_name required' });
    }
    if (worktree !== undefined && !isValidWorktreeName(worktree)) {
      return reply.code(400).send({ error: 'invalid worktree' });
    }
    return relay(reply, 'rename-automation', () =>
      gitops!.renameAutomation({
        bp,
        old_name,
        new_name,
        ...(worktree ? { worktree } : {}),
      }),
    );
  });
}

File: bitswan-gitops/app/routes/templates.py
"""
Automation-template gallery + scaffold endpoints.

Mirror of the dashboard server's old `routes/templates.ts`. The dashboard now
forwards `GET /api/templates` and `POST /api/automations/from-template` to
these endpoints; gitops is the sole writer to `/workspace-repo`.
"""

import logging
import os
import re

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.dependencies import get_automation_service
from app.event_broadcaster import event_broadcaster
from app.services import template_service
from app.services.automation_service import AutomationService

logger = logging.getLogger(__name__)

router = APIRouter(tags=["templates"])

# Same shape as the dashboard server uses for BP / worktree names.
_BP_NAME_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_.-]*$")
_WORKTREE_NAME_RE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9\-]*$")


def _workspace_root() -> str:
    return os.environ.get("BITSWAN_WORKSPACE_REPO_DIR", "/workspace-repo")


@router.get("/templates/")
async def list_templates() -> dict:
    """Return the merged `{templates, groups}` listing."""
    return template_service.discover_templates(_workspace_root())


class CreateFromTemplateRequest(BaseModel):
    template_id: str | None = None
    group_id: str | None = None
    name: str | None = None
    bp: str
    worktree: str | None = None


@router.post("/automations/from-template")
async def create_from_template(
    body: CreateFromTemplateRequest,
    automation_service: AutomationService = Depends(get_automation_service),
) -> dict:
    """Copy a template (or every automation in a group) into the BP directory,
    inject a UUID into each new `automation.toml`, commit, and broadcast the
    fresh `automations` snapshot.
    """
    if not body.bp or not _BP_NAME_RE.match(body.bp):
        raise HTTPException(status_code=400, detail="Invalid bp")
    if body.worktree is not None and not _WORKTREE_NAME_RE.match(body.worktree):
        raise HTTPException(status_code=400, detail="Invalid worktree name")
    if bool(body.template_id) == bool(body.group_id):
        raise HTTPException(
            status_code=400,
            detail="Exactly one of template_id / group_id is required",
        )

    try:
        result = await template_service.create_automation_from_template(
            workspace_root=_workspace_root(),
            bp=body.bp,
            template_id=body.template_id,
            group_id=body.group_id,
            name=body.name,
            worktree=body.worktree,
        )
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.exception("create_automation_from_template failed")
        raise HTTPException(status_code=500, detail=str(e))

    await _broadcast_automations(automation_service, body.worktree)
    return result


async def _broadcast_automations(
    automation_service: AutomationService, worktree: str | None
) -> None:
    """Refresh the affected scope's cache and broadcast the fresh automations
    snapshot so the dashboard updates without waiting for the FS watcher."""
    try:
        await automation_service.refresh(worktree)
        automations = await automation_service.get_automations()
        data = [
            a.model_dump(mode="json") if hasattr(a, "model_dump") else a
            for a in automations
        ]
        await event_broadcaster.broadcast("automations", data)
    except Exception:
        logger.exception("Failed to broadcast automations after change")


# ── Frontend / worker scaffolding ────────────────────────────────────────────
#
# Stage 1.5: business processes are built from frontends (exposed through
# Bailey) and worker containers (private backends). These endpoints scaffold
# them directly from the baked templates — no gallery picker. There is exactly
# one kind of frontend; workers have a `type` (only "go" is wired today,
# mapped to the `go-worker` template; more types will slot in here).

FRONTEND_TEMPLATE_ID = "frontend"
WORKER_TEMPLATE_BY_TYPE = {"go": "go-worker", "fastapi": "fastapi-worker"}


class AddFrontendRequest(BaseModel):
    bp: str
    name: str
    worktree: str | None = None


class AddWorkerRequest(BaseModel):
    bp: str
    name: str
    type: str = "go"
    worktree: str | None = None


class RenameAutomationRequest(BaseModel):
    bp: str
    old_name: str
    new_name: str
    worktree: str | None = None


def _validate_scope(bp: str, worktree: str | None) -> None:
    if not bp or not _BP_NAME_RE.match(bp):
        raise HTTPException(status_code=400, detail="Invalid bp")
    if worktree is not None and not _WORKTREE_NAME_RE.match(worktree):
        raise HTTPException(status_code=400, detail="Invalid worktree name")


async def _scaffold_from_template(
    automation_service: AutomationService,
    *,
    bp: str,
    template_id: str,
    name: str,
    worktree: str | None,
) -> dict:
    try:
        result = await template_service.create_automation_from_template(
            workspace_root=_workspace_root(),
            bp=bp,
            template_id=template_id,
            name=name,
            worktree=worktree,
        )
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.exception("scaffold from template %s failed", template_id)
        raise HTTPException(status_code=500, detail=str(e))
    await _broadcast_automations(automation_service, worktree)
    return result


@router.post("/automations/frontend")
async def add_frontend(
    body: AddFrontendRequest,
    automation_service: AutomationService = Depends(get_automation_service),
) -> dict:
    """Add a frontend (the only kind) to a business process."""
    _validate_scope(body.bp, body.worktree)
    return await _scaffold_from_template(
        automation_service,
        bp=body.bp,
        template_id=FRONTEND_TEMPLATE_ID,
        name=body.name,
        worktree=body.worktree,
    )


@router.post("/automations/worker")
async def add_worker(
    body: AddWorkerRequest,
    automation_service: AutomationService = Depends(get_automation_service),
) -> dict:
    """Add a worker container of the given type to a business process."""
    _validate_scope(body.bp, body.worktree)
    template_id = WORKER_TEMPLATE_BY_TYPE.get(body.type)
    if not template_id:
        raise HTTPException(
            status_code=400,
            detail=f"Unknown worker type {body.type!r}; supported: "
            + ", ".join(sorted(WORKER_TEMPLATE_BY_TYPE)),
        )
    return await _scaffold_from_template(
        automation_service,
        bp=body.bp,
        template_id=template_id,
        name=body.name,
        worktree=body.worktree,
    )


@router.post("/automations/rename")
async def rename_automation_route(
    body: RenameAutomationRequest,
    automation_service: AutomationService = Depends(get_automation_service),
) -> dict:
    """Rename a frontend or worker within a business process."""
    _validate_scope(body.bp, body.worktree)
    try:
        result = await template_service.rename_automation(
            workspace_root=_workspace_root(),
            bp=body.bp,
            old_name=body.old_name,
            new_name=body.new_name,
            worktree=body.worktree,
        )
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.exception("rename_automation failed")
        raise HTTPException(status_code=500, detail=str(e))
    await _broadcast_automations(automation_service, body.worktree)
    return result
